packages/langium/src/syntax-tree.ts

```diff
@@ -240,3 +240,23 @@ export interface RootCstNode extends CompositeCstNode {
 export function isRootCstNode(node: unknown): node is RootCstNode {
     return isCompositeCstNode(node) && 'fullText' in node;
 }
+
+//Manipulates a type to have only properties whose value is of a certain type.
+type ExtractKeysOfValueType<T, K> = { [I in keyof T]: T[I] extends K ? I : never }[keyof T];
+
+/**
+ * Returns the property names (!) of an AstNode that are cross-references.
+ * Meant to be used during cross-reference resolution in combination with `assertUnreachable(context.property)`.
+ */
+export type CrossReferencesOfAstNodeType<N extends AstNode> = (
+    ExtractKeysOfValueType<N, Reference|undefined>
+    | ExtractKeysOfValueType<N, Array<Reference|undefined>|undefined>
+) & Record<string, never>;
```

Comment:
When I apply this to the `ParserRule` type of the grammar language, I get

```
type Bar = ("hiddenTokens" & Record<string, never>) | ("returnType" & Record<string, never>)
```

That looks strange.

Reply:
The reason is the following listing error:
<img width="1057" alt="Bildschirmfoto 2023-03-01 um 09 00 36" src="https://user-images.githubusercontent.com/1525964/222079212-ead170ab-57fc-4978-a5a7-8ef17249a332.png">
But I will use `{}` again and suppress the rule, then it works the best.